src/api-service/__app__/timer_retention/__init__.py

```diff
@@ -0,0 +1,97 @@
+import datetime
+import logging
+
+import azure.functions as func
+from onefuzztypes.enums import JobState, TaskState
+
+from ..onefuzzlib.events import get_events
+from ..onefuzzlib.jobs import Job
+from ..onefuzzlib.notifications.main import Notification
+from ..onefuzzlib.repro import Repro
+from ..onefuzzlib.tasks.main import Task
+
+RETENTION_POLICY = datetime.timedelta(minutes=(5))
+
+
+def main(mytimer1: func.TimerRequest, dashboard: func.Out[str]) -> None:  # noqa: F841
+
+    now = datetime.datetime.now(tz=datetime.timezone.utc)
+
+    time_retained = now - RETENTION_POLICY
+
+    time_filter = f"Timestamp lt datetime'{time_retained.isoformat()}'"
+
+    # You have to do notification before task,
+    # because editing the upn for tasks will change the timestamp
+    for notification in Notification.search(raw_unchecked_filter=time_filter):
+        logging.info("Retention Timer Notification Search")
+        logging.info(
+            "Found notification %s older than 18 months. Checking related tasks.",
+            notification.notification_id,
+        )
+        container = notification.container
+        timestamp_list = []
+        for task in Task.search():
+            container_str = str(task.config.containers)

```

Comment:
Unless you have a strong reason, you should let the type system help you.  Casting will hide errors from you.  Consider a notification container of `foo` and a task with crashes container of `food`.

The code:
```python
if container in container_str:
```
is effectively:
```python
if "foo" in "[TaskContainers(type=<ContainerType.crashes: 'crashes'>, name='food')]":
```
This evaluates to True, but isn't what you want.

Consider TaskConfig.containers:
https://github.com/microsoft/onefuzz/blob/db66a1d3c5d5e44e3780c8a1cc3f349e31f680ce/src/pytypes/onefuzztypes/models.py#L192

Consider TaskContainers.name
https://github.com/microsoft/onefuzz/blob/db66a1d3c5d5e44e3780c8a1cc3f349e31f680ce/src/pytypes/onefuzztypes/models.py#L181-L183

You can have a typed list and use that:
```python
task_containers = [x.name for x in task.config.containers]
if container in task_containers:
```

This would effectively be:
```python
if Container("foo") in [Container("food")]:
```

Reply:
This is good feedback. I'll address appropriately.